viewing_party/party.py

```diff
@@ -1,23 +1,116 @@
+import statistics
+from statistics import mode
+
 # ------------- WAVE 1 --------------------
 
 def create_movie(title, genre, rating):
-    pass
+    new_movie = {}
+    if None in (title, genre, rating):
+        return None
+    else:
+        new_movie["title"] = title
+        new_movie["genre"] = genre
+        new_movie["rating"] = rating
+        return new_movie
+
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+    return user_data
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
+
+def watch_movie(user_data, title):
+    for movie in user_data["watchlist"]:
+        if title != movie["title"]:
+            continue
+        elif movie["title"] == title:
+            user_data["watched"].append(movie)
+            user_data["watchlist"].remove(movie)
+            return user_data
+        
+    return user_data
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    if not len(user_data["watched"]):
+        return 0.0
+    ratings = [movie["rating"] for movie in user_data["watched"]]
+    return statistics.mean(ratings)
+
+def get_most_watched_genre(user_data):
+    if not len(user_data["watched"]):
+        return None
+    genres = [movie["genre"]for movie in user_data["watched"]]
+    return mode(genres)
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    user_movies = get_user_movies(user_data)
+    friend_movies = get_friends_movies(user_data)
+    return [movie for movie in user_movies if movie not in friend_movies]
+
+def get_friends_unique_watched(user_data):
+    user_movies = get_user_movies(user_data)
+    friend_movies = get_friends_movies(user_data)
+    unique_movies = []
+    for movie in friend_movies:
+        if movie not in user_movies and movie not in unique_movies:

```

Comment:
Since we're using the `in` operator here with lists, we will encounter linear time complexity nested within a for loop which would lead us to an overall quadratic time complexity for this method. We could take advantage of constant look up time that `in` has with sets though. 

What if user_movies and friend_movies were sets with just movie titles as strings in them? Then you could have:
```python
for movie in friend_movies:
    if movie["title"] not in user_movies and movie["title"] not in unique_movies: 
        unique_movies.append(movie)
return unique_movies
```

which would bring us to linear time complexity.